### PR TITLE
Add a Linux audit event publisher

### PR DESCRIPTION
**Pull request by theopolis**
This mostly adds support for libaudit, the kernel/system auditing framework. commonly called `auditd` to osquery as an event publisher. As far as I can tell (have researched) the most performant way to support auditing is by requesting to become the `auditd`-like userland process and interpreting the "reply" data from netlink directly. This avoid writing to disk and most of the gotchas around log/data parsing and multi-event correlation. 

Enabling audit support in osquery will blind `auditd` and any reliance on audit logs. As such, support is by default disabled. The publisher will still start and run, but will not manipulate state or affect any `auditd` daemon.

### CODE REVIEW

File: osquery/tables/events/linux/process_events.cpp [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+}
+
+Status ProcessEventSubscriber::Callback(const AuditEventContextRef& ec,
+                                        const void* user_data) {
+  // Define some acceptable transitions outside of the default state.
+  bool acceptable = (ec->type == STATE_PATH && state_ == STATE_EXECVE);
+  acceptable |= (ec->type == STATE_PATH && state_ == STATE_CWD);
+  if (ec->type != state_ && !acceptable) {
+    // This is an out-of-order event, drop it.
+    state_ = STATE_SYSCALL;
```
Comment (theopolis) [inline osquery/tables/events/linux/process_events.cpp]: This should also reset `::row_` data.

File: osquery/tables/events/linux/process_events.cpp [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+  return (s[0] == '"') ? s.substr(1, s.size() - 2) : boost::algorithm::unhex(s);
+}
+
+REGISTER(ProcessEventSubscriber, "event_subscriber", "process_events");
+
+Status ProcessEventSubscriber::init() {
+  auto sc = createSubscriptionContext();
+
+  // Monitor for execve syscalls.
+  sc->rules.push_back({59, ""});
```
Comment (theopolis) [inline osquery/tables/events/linux/process_events.cpp]: Use the define above for more clarity.

File: osquery/events/linux/audit.cpp [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+  bool handle_reply = false;
+  while (true) {
+    handle_reply = false;
+
+    // Request a reply in a non-blocking mode.
+    // This allows the publisher's run loop to periodically request an audit
+    // status update. These updates can check for other processes attempting to
+    // gain control over the audit sink.
+    // This non-blocking also allows faster receipt of multi-message events.
+    result = audit_get_reply(handle_, &reply_, GET_REPLY_NONBLOCKING, 0);
```
Comment (theopolis) [inline osquery/events/linux/audit.cpp]: It's not clear who owns the member of the `reply` structure's member raw pointers? Valgrind plus a python script launching 1000 processes does not show any direct-loss of member. A review of Ubuntu's `auditd` and patches show no obvious freeing. Either way, more investigation is needed.

File: osquery/tables/events/linux/process_events.cpp [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+  // Drop events if they are encountered outside of the expected state.
+  sc->types = {AUDIT_SYSCALL, AUDIT_EXECVE, AUDIT_CWD, AUDIT_PATH};
+  subscribe(&ProcessEventSubscriber::Callback, sc, nullptr);
+
+  return Status(0, "OK");
+}
+
+Status ProcessEventSubscriber::Callback(const AuditEventContextRef& ec,
+                                        const void* user_data) {
+  // Define some acceptable transitions outside of the default state.
```
Comment (theopolis) [inline osquery/tables/events/linux/process_events.cpp]: This method should be broken out into a testable handler, OR, subsequent test runs could inspect the RocksDB store for verification. An inline method returning status and NOT reseting the manipulated `row_` data would be ideal.

File: osquery/events/linux/audit.cpp [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+  }
+}
+
+inline void handleConfigChange(const struct audit_reply& reply) {
+  // Another daemon may have taken control.
+}
+
+inline bool handleReply(const struct audit_reply& reply,
+                        AuditEventContextRef& ec) {
+  // Build an event context around this reply.
```
Comment (theopolis) [inline osquery/events/linux/audit.cpp]: This method is a candidate for testing.

**Comment by javuto** [comment]
ok to test

**Comment by javuto** [comment]
ok to test

**Comment by theopolis** [comment]
This needs:
* a rebase
* a way to add a subscriber without a dependency on syscall or filter. Or a way to make a non-config-modifying subscription

**Pull request merged**